onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_api_helper.cc

```diff
@@ -0,0 +1,157 @@
+
+  int64_t devices_feature_level = -1;
+
+  for (const auto* device_handle : device_handles) {
+    int64_t curr_device_feature_level = 0;
+    if (nnapi->ANeuralNetworksDevice_getFeatureLevel(device_handle, &curr_device_feature_level) != ANEURALNETWORKS_NO_ERROR) {
+      continue;
+    }
+
+    // we want to op run on the device with the highest feature level so we can support more ops.
+    // and we don't care which device runs them.
+    devices_feature_level = std::max(curr_device_feature_level, devices_feature_level);
+  }
+
+  // nnapi-cpu has the feature 1000
+  if ((devices_feature_level > 0) && (devices_feature_level < target_feature_level)) {
+    LOGS_DEFAULT(INFO) << "Changing NNAPI Feature Level " << target_feature_level
+                       << " to supported by target devices: " << devices_feature_level;
+
+    target_feature_level = static_cast<int32_t>(devices_feature_level);
+  }
+  return target_feature_level;
+}
+
+// get all target devices which satisfy the target_device_option
+// we will always put CPU device at the end if cpu is enabled
+Status GetTargetDevices(const ::NnApi* nnapi, TargetDeviceOption target_device_option,
+                        std::vector<ANeuralNetworksDevice*>& nnapi_target_devices, std::string& nnapi_target_devices_detail) {
+  // GetTargetDevices is only supported when NNAPI runtime feature level >= ANEURALNETWORKS_FEATURE_LEVEL_3
+  if (GetNNAPIRuntimeFeatureLevel(nnapi) < ANEURALNETWORKS_FEATURE_LEVEL_3)
+    return Status::OK();
+
+  const std::string nnapi_cpu("nnapi-reference");
+  uint32_t num_devices = 0;
+  RETURN_STATUS_ON_ERROR_WITH_NOTE(
+      nnapi->ANeuralNetworks_getDeviceCount(&num_devices), "Getting count of available devices");
+
+  int32_t cpu_index = -1;
+  for (uint32_t i = 0; i < num_devices; i++) {
+    ANeuralNetworksDevice* device = nullptr;
+    const char* device_name = nullptr;
+    int32_t device_type = 0;
+    RETURN_STATUS_ON_ERROR_WITH_NOTE(
+        nnapi->ANeuralNetworks_getDevice(i, &device), "Getting " + std::to_string(i) + "th device");
+
+    RETURN_STATUS_ON_ERROR_WITH_NOTE(nnapi->ANeuralNetworksDevice_getName(device, &device_name),
+                                     "Getting " + std::to_string(i) + "th device's name");
+
+    RETURN_STATUS_ON_ERROR_WITH_NOTE(nnapi->ANeuralNetworksDevice_getType(device, &device_type),
+                                     "Getting " + std::to_string(i) + "th device's type");
+    bool device_is_cpu = nnapi_cpu == device_name;
+    if ((target_device_option == TargetDeviceOption::CPU_DISABLED) && device_is_cpu) {
+      continue;
+    }
+
+    if (device_is_cpu) {
+      cpu_index = int32_t(nnapi_target_devices.size());
+    }
+    nnapi_target_devices.push_back(device);
+    const auto device_detail = MakeString("[Name: [", device_name, "], Type [", device_type, "]], ");
+    nnapi_target_devices_detail += device_detail;
+  }
+
+  // put CPU device at the end
```

Comment:
nit: explaining 'why' we're doing this would make the comment a more helpful

Reply:
Done